### ISSUE TITLE
Calling a function with no/bad network connection causes exception that cannot be caught

### ISSUE THREAD

**Issue by lexicalninja** [firebase/firebase-android-sdk#1194]
### [REQUIRED] Step 2: Describe your environment

  * Android Studio version: 3.5.3
  * Firebase Component: functions (Database, Firestore, Storage, Functions, etc)
  * Component version: firebase-analytics:17.2.2
    firebase-auth:19.2.0
    firebase-firestore-ktx:21.4.0
    firebase-functions-ktx:19.0.2

### [REQUIRED] Step 3: Describe the problem
When a user's device calls a Firestore Function without any or with bad network connection the app crashes.

The task completed listener is not holding the error and I cannot catch the error after it happens with a try catch block. 
#### Steps to reproduce:
call function with wifi and data turned off

Unfortunately, this is made harder to work around because it is not taking place in a thread/context that we can access the network status, so we need to have some way of catching the error and the SDK seems to be doing something odd with the exception
#### Relevant Code:

```kotlin
if (loggedIn) {
    profileSync.postValue(true)
    try {
        functions.getHttpsCallable("getProfile").call().addOnCompleteListener { task ->
            task.result?.takeIf { task.isSuccessful }?.apply {
                profile = Profile.fromHashMap((task.result?.data as? HashMap<String, Any>))
            } ?: apply {
                Crashlytics.logException(task.exception)
                next.invoke(NETWORK_ERROR)
            }
            profilesRef.document(currentUser!!.uid).collection("sessions")
                    .orderBy("workoutDate", Query.Direction.DESCENDING)
                    .get().addOnCompleteListener { sessionTask ->
                        if (sessionTask.isSuccessful) {
                            sessions.clear()
                            val allSessions = mutableListOf<Session>()
                            sessionTask.result?.apply {
                                this.documents.forEach {
                                    allSessions.add(it.toObject(Session::class.java)!!.withId(it.id))
                                }
                                sessions = allSessions.filter {
                                    it.workoutDate > DateTime.now().minusDays(500).millis.toDouble()
                                }.toMutableList()
                            }
                        } else {
                            Crashlytics.logException(sessionTask.exception)
                        }
                        profilesRef.document(currentUser!!.uid).collection("customWorkouts")
                                .get().addOnCompleteListener {
                                    if (it.isSuccessful) {
                                        customWorkouts.clear()
                                        it.result?.documents?.forEach { doc ->
                                            customWorkouts.add(doc.toObject(Workout::class.java)!!.withId(doc.id))
                                        }
                                    } else {
                                        Crashlytics.logException(it.exception)
                                    }
                                    next.invoke(SUCCESS)
                                    profileSync.postValue(false)
                                }
                    }
        }
    } catch (e: RuntimeExecutionException) {
        Log.d("Kinetic Account", e.localizedMessage)
    }
}
```

here is the Crashlytics output
```kotlin
Fatal Exception: com.google.android.gms.tasks.RuntimeExecutionException: com.google.firebase.functions.FirebaseFunctionsException: Value for argument "data" is not a valid Firestore document. Input is not a plain JavaScript object.
       at com.google.android.gms.tasks.zzu.getResult + 15(:15)
       at com.kinetic.fit.data.firestore.KineticAccount$syncProfile$1.onComplete + 170(KineticAccount.kt:170)
       at com.google.android.gms.tasks.zzj.run + 4(:4)
       at android.os.Handler.handleCallback + 873(Handler.java:873)
       at android.os.Handler.dispatchMessage + 99(Handler.java:99)
       at android.os.Looper.loop + 210(Looper.java:210)
       at android.app.ActivityThread.main + 7124(ActivityThread.java:7124)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run + 493(RuntimeInit.java:493)
       at com.android.internal.os.ZygoteInit.main + 898(ZygoteInit.java:898)

Caused by com.google.firebase.functions.FirebaseFunctionsException: Value for argument "data" is not a valid Firestore document. Input is not a plain JavaScript object.
       at com.google.firebase.functions.FirebaseFunctions$2.onResponse + 281(com.google.firebase:firebase-functions@@19.0.1:281)
       at okhttp3.RealCall$AsyncCall.execute + 206(RealCall.java:206)
       at okhttp3.internal.NamedRunnable.run + 32(NamedRunnable.java:32)
       at java.util.concurrent.ThreadPoolExecutor.runWorker + 1167(ThreadPoolExecutor.java:1167)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run + 641(ThreadPoolExecutor.java:641)
       at java.lang.Thread.run + 764(Thread.java:764)

OkHttp us-central1-kinetic-fit.cloudfunctions.net
       at com.android.org.conscrypt.NativeCrypto.SSL_read(NativeCrypto.java)
       at com.android.org.conscrypt.NativeSsl.read + 399(NativeSsl.java:399)
       at com.android.org.conscrypt.ConscryptFileDescriptorSocket$SSLInputStream.read + 546(ConscryptFileDescriptorSocket.java:546)
       at okio.Okio$2.read + 140(Okio.java:140)
       at okio.AsyncTimeout$2.read + 237(AsyncTimeout.java:237)
       at okio.RealBufferedSource.request + 68(RealBufferedSource.java:68)
       at okio.RealBufferedSource.require + 61(RealBufferedSource.java:61)
       at okhttp3.internal.http2.Http2Reader.nextFrame + 95(Http2Reader.java:95)
       at okhttp3.internal.http2.Http2Connection$ReaderRunnable.execute + 610(Http2Connection.java:610)
       at okhttp3.internal.NamedRunnable.run + 32(NamedRunnable.java:32)
       at java.lang.Thread.run + 764(Thread.java:764)
```

Please let me know if you need any more information for this. We are currently trying for a work around, but being able to catch the error in the task as a failure or consume it in a try catch is preferable. 

Thanks in advance.

**Comment by aguatno** [comment]
Hi @lexicalninja currently, there’s no similar reported issues on our end. I’ve tried to call Firestore Function without network connection, but I'm not experiencing this issue.

If the issue persists, I would like to have a better understanding on how you reproduce this issue, could you provide the following information for my simulation and to discuss this with our team if necessary:

- Detailed steps to reproduce the issue
- A minimal, complete, and verifiable example that we can replicate on our end (without any 3rd party library/tools) 
- Expected vs Actual result
- Any information that you might find helpful for us.

**Comment by samtstern** [comment]
@lexicalninja this error message stands out to me:
```
Fatal Exception: com.google.android.gms.tasks.RuntimeExecutionException: com.google.firebase.functions.FirebaseFunctionsException: Value for argument "data" is not a valid Firestore document. Input is not a plain JavaScript object.
```

I think that means the error is actually inside your Cloud Functions code, not your Android code.  It sounds like you're doing a Firestore `set()` operation in your function but the data is not valid  This causes the function to fail.

Then there's an issue in this part of your code:
```kotlin
functions.getHttpsCallable("getProfile").call().addOnCompleteListener { task ->
            task.result?
```

It looks like you're using `task.result` to check if the task succeeded or not.  However it's not safe to call `Task#getResult()` without knowing if the task succeeded or you'll get a `RuntimeExecutionException`:
https://developers.google.com/android/reference/com/google/android/gms/tasks/Task.html#getResult()

Before calling `.result` you need to check if the `Task` is successful.

So to summarize:

1. Look for errors in your Cloud Functions logs and see if you can figure out where you may be sending bad data to Firestore.
2. In your Android app, check if the Task is successful before trying to get the result.

The combination of those two things should fix your issue!

**Comment by lexicalninja** [comment]
alright, thanks for the help. 

I looked through the functions logs before reporting this, but it wasn't logging the issue. 

Somewhere along the way an aysnc call in the functions code dropped its await modifier, so it was trying to set the whole promise to the document. 

I am still a little confused as to how the try catch couldn't consume the error but I don't feel like that is anything that will be an issue if we check for success first. 

I appreciate the prompt response and assistance.
